common/src/main/java/com/terraformersmc/cinderscapes/mixin/MixinServerWorld.java

```diff
@@ -36,33 +38,38 @@ protected MixinServerWorld(MutableWorldProperties properties, RegistryKey<World>
      */
     @Inject(method="tickIceAndSnow", at = @At(value = "HEAD"), locals = LocalCapture.NO_CAPTURE)
     private void cinderscapes$tickAsh(BlockPos tickPos, CallbackInfo ci) {
+        // Ashy shoals only exists in the nether, why do this iteration on any other dimension
+        if (!getDimensionEntry().matchesKey(DimensionTypes.THE_NETHER)) return;
         if (CinderscapesConfig.INSTANCE.enableAshFall) {
             BlockPos pos = tickPos.mutableCopy();
-            BlockState state = getBlockState(pos);
-            RegistryEntry<Biome> biome = this.getBiome(pos);
 
-            while (pos.getY() < 127 && !(
-                    biome.matchesKey(CinderscapesBiomes.ASHY_SHOALS) &&
-                    state.isSideSolidFullSquare(this, pos, Direction.UP) &&
-                    blockAbove(pos).isIn(CinderscapesBlockTags.ASH_PERMEABLE) &&
-                    this.getBlockState(pos.up()).isAir() &&
-                    CinderscapesBlocks.ASH.getDefaultState().canPlaceAt(this, pos.up()))) {
-                pos = pos.up();
-                state = getBlockState(pos);
-                biome = this.getBiome(pos);
-            }
-
-            if (pos.getY() < 127) {
-                this.setBlockState(pos.up(), CinderscapesBlocks.ASH.getDefaultState());
+            for (; pos.getY() < 127; pos.setY(pos.getY() + 1)) {
+                BlockPos up = pos.up();
+                if (!canPlace(up)) {
+                    continue;
+                } else if (!this.getBiome(up).matchesKey(CinderscapesBiomes.ASHY_SHOALS)) {
+                    continue;
+                }
+                this.setBlockState(up, CinderscapesBlocks.ASH.getDefaultState());
+                break;
```

Comment:
Rather than using two equivalent blocks, could the conditions here be inverted to form a single if statement without using `continue`?

Reply:
I think the guard clause arrangement is pretty clean as-is, but of course if a better logic layout is possible it should be used.

Reply:
That code is equivalent to:

```suggestion
                if (canPlace(up) && this.getBiome(up).matchesKey(CinderscapesBiomes.ASHY_SHOALS)) {
                    this.setBlockState(up, CinderscapesBlocks.ASH.getDefaultState());
                }
```